### ISSUE TITLE
Using custom functions in TypeQL queries

### ISSUE THREAD

**Issue by haikalpribadi** [vaticle/typedb#5528]
## Problem to Solve

Given the following `match` clause:
```
match $x isa declaration, has expense $e;
```

We would like to apply a custom function on `$e`, such as: 
```
$i = roundInt($e);
```

The custom functions we would like to perform are:
- `roundInt($e)`
- `ceil($e)`
- `floor($e)`
- `min($e, $f)`
- `max($e, $f)`

## Current Workaround

At the moment, users have to compute these operations manually and store them into the database.

## Proposed Solution

Similar to arithmetic operations (#5521), we introduce a special type of `statement` to perform custom functions; let's say _"functional statements"_. This would allow us to write `pattern`s (in `match` clauses and `rule` bodies) such as:
```
$x isa declaration, has expense $e;
$i = roundInt($e);
```

## Additional Information

As a first iteration of the solution, we can prohibit variables that contain results of custom functions to be used further for comparisons.

**Comment by brettforbes** [comment]
Can this include log2 and log10?
Log2 brings the possibility of calculating entropy and is very important

**Comment by flyingsilverfin** [comment]
The initial set has been implemented in #6654 , please open new feature requests for new functionalities
